server/tso/allocator_manager.go

```diff
@@ -0,0 +1,197 @@
+	"sync"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/election"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/embed"
+	"go.uber.org/zap"
+)
+
+type allocatorGroup struct {
+	dcLocation string
+	// allocator's parent ctx and cancel function, which is to
+	// control the allocator's behavior in AllocatorDaemon and
+	// pass the cancel signal to its parent as soon as possible
+	// since it is critical to let parent goroutine know whether
+	// the allocator is still able to work well.
+	parentCtx    context.Context
+	parentCancel context.CancelFunc
+	// For the Global TSO Allocator, leadership is a PD leader's
+	// leadership, and for the Local TSO Allocator, leadership
+	// is a DC-level certificate to allow an allocator to generate
+	// TSO for local transactions in its DC.
+	leadership *election.Leadership
+	allocator  Allocator
+	// the flag indicates whether this allocator is initialized
+	isInitialized bool
+}
+
+// AllocatorManager is used to manage the TSO Allocators a PD server holds.
+// It is in charge of maintaining TSO allocators' leadership, checking election
+// priority, and forwarding TSO allocation requests to correct TSO Allocators.
+type AllocatorManager struct {
+	sync.RWMutex
+	wg sync.WaitGroup
+	// There are two kinds of TSO Allocators:
+	//   1. Global TSO Allocator, as a global single point to allocate
+	//      TSO for global transactions, such as cross-region cases.
+	//   2. Local TSO Allocator, servers for DC-level transactions.
+	// dc-location/global (string) -> TSO Allocator
+	allocatorGroups map[string]*allocatorGroup
+	// etcd and its client
+	etcd   *embed.Etcd
+	client *clientv3.Client
+	// tso config
+	rootPath      string
+	saveInterval  time.Duration
+	maxResetTSGap func() time.Duration
+}
+
+// NewAllocatorManager creates a new TSO Allocator Manager.
+func NewAllocatorManager(etcd *embed.Etcd, client *clientv3.Client, rootPath string, saveInterval time.Duration, maxResetTSGap func() time.Duration) *AllocatorManager {
+	allocatorManager := &AllocatorManager{
+		allocatorGroups: make(map[string]*allocatorGroup),
+		etcd:            etcd,
+		client:          client,
+		rootPath:        rootPath,
+		saveInterval:    saveInterval,
+		maxResetTSGap:   maxResetTSGap,
+	}
+	return allocatorManager
+}
+
+func (am *AllocatorManager) getAllocatorPath(dcLocation string) string {
+	// For backward compatibility, the global timestamp's store path will still use the old one
+	if dcLocation == "global" {
+		return am.rootPath
+	}
+	return path.Join(am.rootPath, dcLocation)
+}
+
+// SetUpAllocator is used to set up an allocator, which will initialize the allocator and put it into allocator daemon.
+func (am *AllocatorManager) SetUpAllocator(parentCtx context.Context, parentCancel context.CancelFunc, dcLocation string, leadership *election.Leadership) error {
+	am.Lock()
+	defer am.Unlock()
+	switch dcLocation {
+	case "global":
```

Comment:
How about using a constant for "global"?

Reply:
What about using a random string to be the value of this constant to prevent it from being duplicated from the user's `dc-location`? Like this:

```go
const globalDCLocation = "GLOBAL_b6b3f51996c7da2d8adcbb44baa39750"
```

Reply:
I think we should prevent "global" from being duplicated by user's input dc-location by adding validation in the api instead of generating a random name which will be misleading for the user in the view.